File: web/src/services/qa.ts
import { request } from './http'
import type {
  QaEvidence,
  QaFeedback,
  QaFeedbackView,
  QaMessageCreate,
  QaMessageList,
  QaScopeUpdate,
  QaSession,
  QaSessionCreate,
  QaSessionList,
  QaTask,
} from '../types/qa'

const segment = encodeURIComponent
export const qaApi = {
  sessions: (page = 1, signal?: AbortSignal) =>
    request<QaSessionList>(`/qa/sessions?page=${page}&page_size=20`, { signal }),
  createSession: (data: QaSessionCreate, signal?: AbortSignal) =>
    request<QaSession>('/qa/sessions', { method: 'POST', data, signal }),
  session: (id: string, signal?: AbortSignal) =>
    request<QaSession>(`/qa/sessions/${segment(id)}`, { signal }),
  updateScope: (id: string, data: QaScopeUpdate, signal?: AbortSignal) =>
    request<QaSession>(`/qa/sessions/${segment(id)}/scope`, { method: 'PATCH', data, signal }),
  messages: (id: string, before?: number, signal?: AbortSignal) => {
    const params = new URLSearchParams({ page_size: '20' })
    if (before !== undefined) params.set('before_sequence', String(before))
    return request<QaMessageList>(`/qa/sessions/${segment(id)}/messages?${params}`, { signal })
  },
  ask: (id: string, data: QaMessageCreate, key: string, signal?: AbortSignal) =>
    request<QaTask>(`/qa/sessions/${segment(id)}/messages`, {
      method: 'POST',
      data,
      idempotencyKey: key,
      signal,
    }),
  task: (id: string, signal?: AbortSignal) =>
    request<QaTask>(`/qa/tasks/${segment(id)}`, { signal }),
  cancel: (id: string, signal?: AbortSignal) =>
    request<QaTask>(`/qa/tasks/${segment(id)}/cancel`, { method: 'POST', signal }),
  evidence: (answerId: string, evidenceId: string, signal?: AbortSignal) =>
    request<QaEvidence>(`/qa/answers/${segment(answerId)}/evidence/${segment(evidenceId)}`, {
      signal,
    }),
  feedback: (answerId: string, data: QaFeedback, signal?: AbortSignal) =>
    request<QaFeedbackView>(`/qa/answers/${segment(answerId)}/feedback`, {
      method: 'POST',
      data,
      signal,
    }),
}

export const qaKeys = {
  sessions: (identity: string | number) => [identity, 'qa', 'sessions'] as const,
  session: (identity: string | number, sessionId: string) =>
    [identity, 'qa', sessionId, 'session'] as const,
  messages: (identity: string | number, sessionId: string) =>
    [identity, 'qa', sessionId, 'messages'] as const,
  task: (identity: string | number, sessionId: string, taskId?: string) =>
    [identity, 'qa', sessionId, 'task', taskId] as const,
  evidence: (identity: string | number, sessionId: string) =>
    [identity, 'qa', sessionId, 'evidence'] as const,
}

File: backend/app/api/v1/routes/qa.py
from fastapi import APIRouter, Depends, Header, Query, Request

from app.core.auth import get_current_actor
from app.models.common import ApiResponse
from app.models.qa import (
    QaFeedbackBody,
    QaFeedbackView,
    QaMessageCreate,
    QaMessageList,
    QaScopeUpdate,
    QaSessionCreate,
    QaSessionList,
    QaSessionView,
    QaTaskView,
)
from app.rag.contracts import DocumentEvidence
from app.services import qa_read, qa_service, task_event_stream

router = APIRouter(prefix="/qa", tags=["knowledge-qa"])


@router.get("/tasks/{task_id}/content-events")
async def content_events(task_id: str, request: Request,
                         last_event_id: str | None = Header(default=None, alias="Last-Event-ID"),
                         actor=Depends(get_current_actor)):
    from app.services.content_event_stream import content_events_endpoint

    return await content_events_endpoint(request, actor=actor, kind="qa", task_id=task_id, last_event_id=last_event_id)


@router.post("/sessions", status_code=201, response_model=ApiResponse[QaSessionView])
async def create_session(body: QaSessionCreate, actor=Depends(get_current_actor)):
    return ApiResponse.success(await qa_service.create_session(actor, body))


@router.get("/sessions", response_model=ApiResponse[QaSessionList])
async def sessions(
    page: int = Query(1, ge=1, le=100000),
    page_size: int = Query(20, ge=1, le=50),
    actor=Depends(get_current_actor),
):
    return ApiResponse.success(
        await qa_read.list_sessions(actor.owner_id, page, page_size)
    )


@router.get("/sessions/{session_id}", response_model=ApiResponse[QaSessionView])
async def session(session_id: str, actor=Depends(get_current_actor)):
    return ApiResponse.success(await qa_read.get_session(actor.owner_id, session_id))


@router.patch("/sessions/{session_id}/scope", response_model=ApiResponse[QaSessionView])
async def scope(session_id: str, body: QaScopeUpdate, actor=Depends(get_current_actor)):
    return ApiResponse.success(await qa_service.update_scope(actor, session_id, body))


@router.get(
    "/sessions/{session_id}/messages", response_model=ApiResponse[QaMessageList]
)
async def messages(
    session_id: str,
    before_sequence: int | None = Query(None, ge=1),
    page_size: int = Query(30, ge=1, le=100),
    actor=Depends(get_current_actor),
):
    return ApiResponse.success(
        await qa_read.get_messages(
            actor.owner_id, session_id, before_sequence, page_size
        )
    )


@router.post(
    "/sessions/{session_id}/messages",
    status_code=202,
    response_model=ApiResponse[QaTaskView],
)
async def create_message(
    session_id: str,
    body: QaMessageCreate,
    actor=Depends(get_current_actor),
    idempotency_key: str = Header(..., min_length=1, max_length=128),
):
    return ApiResponse.success(
        await qa_service.create_message(actor, session_id, body, idempotency_key)
    )


@router.get("/tasks/{task_id}", response_model=ApiResponse[QaTaskView])
async def task(task_id: str, actor=Depends(get_current_actor)):
    return ApiResponse.success(await qa_read.get_task(actor.owner_id, task_id))


@router.get("/tasks/{task_id}/events")
async def task_events(
    task_id: str,
    request: Request,
    last_event_id: str | None = Header(default=None, alias="Last-Event-ID"),
    actor=Depends(get_current_actor),
):
    """任务阶段 SSE 流；授权与快照一致读取见 task_event_stream。"""
    return await task_event_stream.task_events_endpoint(
        request,
        actor=actor,
        kind="qa",
        task_id=task_id,
        last_event_id=last_event_id,
    )


@router.post("/tasks/{task_id}/cancel", response_model=ApiResponse[QaTaskView])
async def cancel(task_id: str, actor=Depends(get_current_actor)):
    return ApiResponse.success(await qa_service.cancel_task(actor.owner_id, task_id))


@router.get(
    "/answers/{answer_id}/evidence/{evidence_id}",
    response_model=ApiResponse[DocumentEvidence],
)
async def evidence(answer_id: str, evidence_id: str, actor=Depends(get_current_actor)):
    return ApiResponse.success(
        await qa_read.read_evidence(actor.owner_id, answer_id, evidence_id)
    )


@router.post(
    "/answers/{answer_id}/feedback", response_model=ApiResponse[QaFeedbackView]
)
async def feedback(
    answer_id: str, body: QaFeedbackBody, actor=Depends(get_current_actor)
):
    return ApiResponse.success(
        await qa_service.save_feedback(actor.owner_id, answer_id, body)
    )
